Performance regression on CouchDB v3 while using custom reduce function

I was trying to upgrade my database installation from CouchDB v2.3.1 to v3.1.1 and experienced about 6 time slower performance on particular view invocations.

## Description

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT give us access or passwords to your actual CouchDB! )

## Steps to Reproduce
To make it easier to reproduce, I created a script which sets up CouchDB 2.3.1 and CouchDB 3.1.1 in docker/podman, generates example data and imports the problematic view.
The test script generates 12000 documents and a single design document including the slow view.
The script can be used to measure response times against the local system.

[2021_04_19-couchdb-test.zip](https://github.com/apache/couchdb/files/6337054/2021_04_19-couchdb-test.zip)

## Show usage
```
$ ./couchdb-test.sh 
Usage: couchdb-test.sh setup|query|cleanup
```

## Initial setup of scenario to reproduce
The following step takes about 13 min to run, the data generation is not very efficient with curl, but I did not want to force to much requirements on the host except of having bash, curl and docker/podman installed.
```
$ ./couchdb-test.sh setup
CouchDB performance regression test script
Initializing clean CouchDB setups
Cleanup containers if present
Error: failed to evict container: "": failed to find container "couchdb2" in state: no container with name or ID couchdb2 found: no such container
Create container
e0077b6673c9751053d4c9613c38c5de0015a283315a3dd4f125a6242de835af
Cleanup containers if present
Error: failed to evict container: "": failed to find container "couchdb3" in state: no container with name or ID couchdb3 found: no such container
Create container
d114d4a2e6eb28f8f322e85083e46aa033701206a13c1789c6beccc8d457f656
Setup CouchDB2
Waiting for CouchDB2 to become available at http://localhost:3002...
CouchDB is up and running
Creating _users
{"ok":true}
Creating _replicator
{"ok":true}
Creating _global_changes
{"ok":true}
Creating demo
{"ok":true}
Setup CouchDB3
Waiting for CouchDB2 to become available at http://localhost:3003...
CouchDB is up and running
Creating _users
{"ok":true}
Creating _replicator
{"ok":true}
Creating _global_changes
{"ok":true}
Creating demo
{"ok":true}
Creating design doc
{"ok":true,"id":"_design/demo","rev":"1-6732411a147da115d79425ba6aeb71de"}
Creating design doc
{"ok":true,"id":"_design/demo","rev":"1-6732411a147da115d79425ba6aeb71de"}
Generating data
inserting docs cycle 1/2000
{"ok":true,"id":"5d933844b700c58fc1b422deb0000687","rev":"1-8353a76e296fb10a71bfb93b7b0ae73f"}
{"ok":true,"id":"5d933844b700c58fc1b422deb0000cf3","rev":"1-60f7912abd240b2433b1ea010f0699c4"}
{"ok":true,"id":"5d933844b700c58fc1b422deb0001801","rev":"1-1a83628882bcb89401df77e7fcc42fd1"}
{"ok":true,"id":"5d933844b700c58fc1b422deb0002000","rev":"1-8a91cdc871ffb7623369ab81d2b6e980"}
{"ok":true,"id":"5d933844b700c58fc1b422deb00023ce","rev":"1-66ac9947820d8f32ca321a91d11b8794"}
...
inserting docs cycle 1999/2000
{"ok":true,"id":"01594d00a265804fd144a3e55575806e","rev":"1-6620f7ce5b499af7486ba404b43fcf6a"}
{"ok":true,"id":"01594d00a265804fd144a3e555758337","rev":"1-6595ca907b8e04dd56f83c56a1ccb5ee"}
{"ok":true,"id":"01594d00a265804fd144a3e5557584f1","rev":"1-31bfdc81ae9ab7c8d5eb16007f479309"}
{"ok":true,"id":"01594d00a265804fd144a3e555758f9e","rev":"1-b3109111048a9a14bce96b1a6a5df642"}
{"ok":true,"id":"01594d00a265804fd144a3e555759188","rev":"1-37fc4bd5012c2e90c13324cc7911d053"}
{"ok":true,"id":"01594d00a265804fd144a3e555759880","rev":"1-7457236eacd2fc5742491a431bccabdb"}
inserting docs cycle 2000/2000
{"ok":true,"id":"01594d00a265804fd144a3e55575a0a7","rev":"1-f7fac9a3f6ce538d67747271336211c4"}
{"ok":true,"id":"01594d00a265804fd144a3e55575ab65","rev":"1-7e3d2b01acbc416b60e6e5be050dd790"}
{"ok":true,"id":"01594d00a265804fd144a3e55575b4de","rev":"1-c959b425a724ad6346f646d8a63da6cd"}
{"ok":true,"id":"01594d00a265804fd144a3e55575c456","rev":"1-d8bd0d00d0b04b52796cfe73d9a80760"}
{"ok":true,"id":"01594d00a265804fd144a3e55575cbcb","rev":"1-24f6bc4fcfdceae7ff2fae47a68fbf42"}
{"ok":true,"id":"01594d00a265804fd144a3e55575d758","rev":"1-b0dfc903b7f27dc98130c8d51931f5d4"}
```
## Query performance
```
$ ./couchdb-test.sh query
CouchDB performance regression test script
Assuming that setup is already complete
================ Query CouchDB 2.3.1 ================
{"couchdb":"Welcome","version":"2.3.1","git_sha":"c298091a4","uuid":"9772359d7167d42054df2c9bca1388c9","features":["pluggable-storage-engines","scheduler"],"vendor":{"name":"The Apache Software Foundation"}}
Query Database views on port 3002
Round: 1/10

real	0m6.258s
user	0m0.013s
sys	0m0.012s
Round: 2/10

real	0m6.526s
user	0m0.014s
sys	0m0.011s
Round: 3/10

real	0m6.477s
user	0m0.010s
sys	0m0.015s
Round: 4/10

real	0m6.156s
user	0m0.011s
sys	0m0.014s
Round: 5/10

real	0m6.185s
user	0m0.015s
sys	0m0.011s
Round: 6/10

real	0m6.082s
user	0m0.015s
sys	0m0.015s
Round: 7/10

real	0m5.932s
user	0m0.012s
sys	0m0.014s
Round: 8/10

real	0m5.594s
user	0m0.013s
sys	0m0.013s
Round: 9/10

real	0m6.258s
user	0m0.012s
sys	0m0.014s
Round: 10/10

real	0m5.796s
user	0m0.009s
sys	0m0.017s
================ Query CouchDB 3.1.1 ================
{"couchdb":"Welcome","version":"3.1.1","git_sha":"ce596c65d","uuid":"5f8a19fe389af901b8044f541b0d3c9f","features":["access-ready","partitioned","pluggable-storage-engines","reshard","scheduler"],"vendor":{"name":"The Apache Software Foundation"}}
Query Database views on port 3003
Round: 1/10

real	0m42.177s
user	0m0.012s
sys	0m0.016s
Round: 2/10

real	0m39.650s
user	0m0.013s
sys	0m0.016s
Round: 3/10

real	0m40.073s
user	0m0.013s
sys	0m0.014s
Round: 4/10

real	0m40.407s
user	0m0.011s
sys	0m0.017s
Round: 5/10

real	0m39.927s
user	0m0.013s
sys	0m0.015s
Round: 6/10

real	0m38.747s
user	0m0.012s
sys	0m0.016s
Round: 7/10

real	0m38.286s
user	0m0.014s
sys	0m0.013s
Round: 8/10

real	0m38.962s
user	0m0.015s
sys	0m0.014s
Round: 9/10

real	0m37.821s
user	0m0.017s
sys	0m0.012s
Round: 10/10

real	0m37.505s
user	0m0.015s
sys	0m0.013s

``` 

## Cleanup
```
$ ./couchdb-test.sh cleanup
CouchDB performance regression test script
CONTAINER ID  IMAGE                            COMMAND               CREATED        STATUS            PORTS                   NAMES
b885253d64ea  docker.io/library/couchdb:3.1.1  /opt/couchdb/bin/...  2 minutes ago  Up 2 minutes ago  0.0.0.0:3003->5984/tcp  couchdb3
908b29168c7e  docker.io/library/couchdb:2.3.1  /opt/couchdb/bin/...  2 minutes ago  Up 2 minutes ago  0.0.0.0:3002->5984/tcp  couchdb2
Removing containers couchdb2 and couchdb3 if present
908b29168c7e35b6a0dda9d2723c54ff972e44b8467ccd05b28421d033c48a9d
b885253d64ea11a40cbb2401d4f9e04d6a34cdffcf375c37eb989d19a4f9db3f
CONTAINER ID  IMAGE   COMMAND  CREATED  STATUS  PORTS   NAMES

$ podman rmi docker.io/library/couchdb:2.3.1 docker.io/library/couchdb:3.1.1
```

## Expected Behaviour
The performance of the view should not decrease, or at least not at that magnitude.

## Your Environment


* CouchDB version used: v2.3.1 and v3.1.1 (official unmodified Container images from https://hub.docker.com/_/couchdb)
* Browser name and version: curl
* Operating system and version: I was running the script on RHEL 8.3 with podman version 2.2.1.

## Additional Context
Used map function:
```
function (doc) {
    if(doc.object === "ainst" && doc.result) {
        var subscription = doc.subscription || doc.tenant;
        var stats = {
            successfulCount : 0,
            failedCount : 0,
            canceledCount : 0,
            executingCount : 0,
            unsuccessfulCount : 0,
            unknownCount : 0,
            totalCount : 1,
            execTimeSum : 0,
            execTimeCount : 0,
            execTimeMin : Infinity,
            execTimeMax : 0
        }

        if (doc.result.status === "successful") {
            stats.successfulCount = 1;

            if (doc.result.executionTime) {
                stats.execTimeSum = doc.result.executionTime;
                stats.execTimeCount = 1;
                stats.execTimeMin = doc.result.executionTime;
                stats.execTimeMax = doc.result.executionTime;
            }
        } else if (doc.result.status === "failed") {
            stats.failedCount = 1;
        } else if (doc.result.status === "canceled") {
            stats.canceledCount = 1;
        } else if (doc.result.status === "executing") {
            stats.executingCount = 1;
        } else if (doc.result.status === "unsuccessful") {
            stats.unsuccessfulCount = 1;
        } else if (doc.result.status === "unknown") {
            stats.unknownCount = 1;
        }

        emit([subscription, doc.aDefId], stats);
    }
}
```

Custom reduce function:
```
function (keys, values, rereduce) {
    var stats = {
        successfulCount : 0,
        failedCount : 0,
        canceledCount : 0,
        executingCount : 0,
        unsuccessfulCount : 0,
        unknownCount : 0,
        totalCount : 0,
        execTimeSum : 0,
        execTimeCount : 0,
        execTimeMin : Infinity,
        execTimeMax : 0
    }

    for (var index in values) {
        stats.successfulCount += values[index].successfulCount;
        stats.failedCount += values[index].failedCount;
        stats.canceledCount += values[index].canceledCount;
        stats.executingCount += values[index].executingCount;
        stats.unsuccessfulCount += values[index].unsuccessfulCount;
        stats.unknownCount += values[index].unknownCount;
        stats.totalCount += values[index].totalCount;
        if (values[index].execTimeCount) {
            stats.execTimeSum += values[index].execTimeSum;
            stats.execTimeCount += values[index].execTimeCount;
            stats.execTimeMin = Math.min(stats.execTimeMin, values[index].execTimeMin);
            stats.execTimeMax = Math.max(stats.execTimeMax, values[index].execTimeMax);
        }
    }

    return stats;
}
```

CouchDB v2.3.1 logs
```
[notice] 2021-04-19T15:15:56.957812Z nonode@nohost <0.8597.3> ea0847955d 127.0.0.1:3002 10.0.2.100 admin GET / 200 ok 1
[info] 2021-04-19T15:15:57.024251Z nonode@nohost <0.8708.3> -------- Starting index update for db: shards/20000000-3fffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:15:57.024263Z nonode@nohost <0.8710.3> -------- Starting index update for db: shards/00000000-1fffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:15:57.027396Z nonode@nohost <0.8714.3> -------- Starting index update for db: shards/80000000-9fffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:15:57.027402Z nonode@nohost <0.8716.3> -------- Starting index update for db: shards/e0000000-ffffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:15:57.027414Z nonode@nohost <0.8717.3> -------- Starting index update for db: shards/a0000000-bfffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:15:57.027510Z nonode@nohost <0.8720.3> -------- Starting index update for db: shards/60000000-7fffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:15:57.033749Z nonode@nohost <0.8726.3> -------- Starting index update for db: shards/c0000000-dfffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:15:57.037107Z nonode@nohost <0.8729.3> -------- Starting index update for db: shards/40000000-5fffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:16:00.545524Z nonode@nohost <0.8714.3> -------- Index update finished for db: shards/80000000-9fffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:16:00.586724Z nonode@nohost <0.8708.3> -------- Index update finished for db: shards/20000000-3fffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:16:00.602767Z nonode@nohost <0.8710.3> -------- Index update finished for db: shards/00000000-1fffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:16:00.621844Z nonode@nohost <0.8720.3> -------- Index update finished for db: shards/60000000-7fffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:16:00.634404Z nonode@nohost <0.8716.3> -------- Index update finished for db: shards/e0000000-ffffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:16:00.639086Z nonode@nohost <0.8729.3> -------- Index update finished for db: shards/40000000-5fffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:16:00.642827Z nonode@nohost <0.8726.3> -------- Index update finished for db: shards/c0000000-dfffffff/demo.1618843969 idx: _design/demo
[info] 2021-04-19T15:16:00.660120Z nonode@nohost <0.8717.3> -------- Index update finished for db: shards/a0000000-bfffffff/demo.1618843969 idx: _design/demo
[notice] 2021-04-19T15:16:04.871512Z nonode@nohost <0.8598.3> 0ff4f8d0e3 127.0.0.1:3002 10.0.2.100 admin GET /demo/_design/demo/_view/slow?reduce=true&group=true 200 ok 7901
[notice] 2021-04-19T15:16:09.628157Z nonode@nohost <0.8904.3> 7c47c98e62 127.0.0.1:3002 10.0.2.100 admin GET /demo/_design/demo/_view/slow?reduce=true&group=true 200 ok 4749
[notice] 2021-04-19T15:16:14.377989Z nonode@nohost <0.9008.3> 15f2ef83f8 127.0.0.1:3002 10.0.2.100 admin GET /demo/_design/demo/_view/slow?reduce=true&group=true 200 ok 4742
[notice] 2021-04-19T15:16:19.214627Z nonode@nohost <0.9086.3> c1231fddeb 127.0.0.1:3002 10.0.2.100 admin GET /demo/_design/demo/_view/slow?reduce=true&group=true 200 ok 4830
```

CouchDB v3.1.1 logs
```
[notice] 2021-04-19T15:16:50.323986Z nonode@nohost <0.16853.3> 105c664a0b 127.0.0.1:3003 10.0.2.100 admin GET / 200 ok 1
[notice] 2021-04-19T15:17:26.534828Z nonode@nohost <0.16854.3> e06cf40a83 127.0.0.1:3003 10.0.2.100 admin GET /demo/_design/demo/_view/slow?reduce=true&group=true 200 ok 36202
[notice] 2021-04-19T15:18:05.599146Z nonode@nohost <0.17182.3> 3db1f4acb4 127.0.0.1:3003 10.0.2.100 admin GET /demo/_design/demo/_view/slow?reduce=true&group=true 200 ok 39056
[notice] 2021-04-19T15:18:44.043762Z nonode@nohost <0.17518.3> 4b7c9097f9 127.0.0.1:3003 10.0.2.100 admin GET /demo/_design/demo/_view/slow?reduce=true&group=true 200 ok 38437
```


It might be related to the change in the JavaScript engine in CouchDB 3.0, but I am just guessing https://blog.couchdb.org/2020/02/26/the-road-to-couchdb-3-0-update-to-javascript-engine/

Thanks in advance for any help
Konrad

Okay, thanks, looks like I need to go with libmozjs-78, if I want to use debian bullseye repos:

```bash
$ podman run --rm -it debian:bullseye-slim
root@e0167e5dead8:/# apt-get update
Get:1 http://deb.debian.org/debian bullseye InRelease [116 kB]
Get:2 http://deb.debian.org/debian-security bullseye-security InRelease [48.4 kB]
Get:3 http://deb.debian.org/debian bullseye-updates InRelease [44.1 kB]
Get:4 http://deb.debian.org/debian bullseye/main amd64 Packages [8184 kB]
Get:5 http://deb.debian.org/debian-security bullseye-security/main amd64 Packages [209 kB]
Get:6 http://deb.debian.org/debian bullseye-updates/main amd64 Packages [14.6 kB]
Fetched 8615 kB in 1s (7371 kB/s)                           
Reading package lists... Done
root@e0167e5dead8:/# apt-cache search libmozjs
libmozjs-78-0 - SpiderMonkey JavaScript library
libmozjs-78-dev - SpiderMonkey JavaScript library - development headers
```

Therefore, I recompiled by doing the following steps (reusing the build container images, just recompile CouchDB), unfortunatly, `make release` fails in that case. I tried again with defaults `./configure && make release` which seem to work.

```bash
$ podman run --name couchdb-fix-3517 -it localhost/apache/couchdbci-debian:buster-erlang-23.3.4.15
jenkins@401db7b11852:/$ git clone --depth=1 -b fix/3517-part7 https://github.com/apache/couchdb.git
fatal: could not create work tree dir 'couchdb': Permission denied
jenkins@401db7b11852:/$ cd ~
jenkins@401db7b11852:~$ git clone --depth=1 -b fix/3517-part7 https://github.com/apache/couchdb.git
Cloning into 'couchdb'...
remote: Enumerating objects: 1508, done.
remote: Counting objects: 100% (1508/1508), done.
remote: Compressing objects: 100% (1309/1309), done.
remote: Total 1508 (delta 289), reused 625 (delta 138), pack-reused 0
Receiving objects: 100% (1508/1508), 4.95 MiB | 18.51 MiB/s, done.
Resolving deltas: 100% (289/289), done.
jenkins@401db7b11852:~$ cd couchdb/
jenkins@401db7b11852:~/couchdb$ git log
commit f6d262aa2bfe5d060cc8798f7d8d00b71f61cfe7 (grafted, HEAD -> fix/3517-part7, origin/fix/3517-part7)
Author: Jan Lehnardt <jan@apache.org>
Date:   Tue Dec 6 12:50:15 2022 +0100

    fix(3517): super-simplistic fix to avoid costly AST transforms when they are not needed
jenkins@401db7b11852:~/couchdb$ ./configure --spidermonkey-version 78
==> configuring couchdb in rel/couchdb.config
Cloning into '/home/jenkins/couchdb/src/rebar'...
remote: Enumerating objects: 8885, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 8885 (delta 4), reused 9 (delta 4), pack-reused 8875
Receiving objects: 100% (8885/8885), 2.16 MiB | 19.44 MiB/s, done.
Resolving deltas: 100% (5705/5705), done.
make: Entering directory '/home/jenkins/couchdb/src/rebar'
./bootstrap
No beam files found.
Recompile: src/rebar
Recompile: src/rebar_abnfc_compiler
Recompile: src/rebar_app_utils
Recompile: src/rebar_appups
Recompile: src/rebar_asn1_compiler
Recompile: src/rebar_base_compiler
Recompile: src/rebar_cleaner
Recompile: src/rebar_config
Recompile: src/rebar_core
Recompile: src/rebar_cover_utils
Recompile: src/rebar_ct
src/rebar_ct.erl:291: Warning: crypto:rand_uniform/2 is deprecated; use rand:uniform/1 instead
Recompile: src/rebar_deps
Recompile: src/rebar_dia_compiler
Recompile: src/rebar_dialyzer
Recompile: src/rebar_edoc
Recompile: src/rebar_erlc_compiler
Recompile: src/rebar_erlydtl_compiler
Recompile: src/rebar_escripter
Recompile: src/rebar_eunit
src/rebar_eunit.erl:282: Warning: crypto:rand_uniform/2 is deprecated; use rand:uniform/1 instead
Recompile: src/rebar_file_utils
Recompile: src/rebar_getopt
Recompile: src/rebar_lfe_compiler
Recompile: src/rebar_log
Recompile: src/rebar_metacmds
Recompile: src/rebar_mustache
Recompile: src/rebar_neotoma_compiler
Recompile: src/rebar_otp_app
Recompile: src/rebar_otp_appup
Recompile: src/rebar_port_compiler
Recompile: src/rebar_proto_compiler
Recompile: src/rebar_proto_gpb_compiler
Recompile: src/rebar_protobuffs_compiler
Recompile: src/rebar_qc
Recompile: src/rebar_rand_compat
Recompile: src/rebar_rel_utils
Recompile: src/rebar_reltool
Recompile: src/rebar_require_vsn
Recompile: src/rebar_shell
Recompile: src/rebar_subdirs
Recompile: src/rebar_templater
Recompile: src/rebar_upgrade
Recompile: src/rebar_utils
Recompile: src/rebar_xref
Recompile: src/rmemo
==> rebar (compile)
==> rebar (escriptize)
Congratulations! You now have a self-contained script called "rebar" in
your current working directory. Place this script anywhere in your path
and you can use rebar to build OTP-compliant apps.
make: Leaving directory '/home/jenkins/couchdb/src/rebar'
make: Entering directory '/home/jenkins/couchdb/src/rebar'
make: Leaving directory '/home/jenkins/couchdb/src/rebar'
Cloning into '/home/jenkins/couchdb/src/rebar3'...
remote: Enumerating objects: 515, done.
remote: Counting objects: 100% (515/515), done.
remote: Compressing objects: 100% (445/445), done.
remote: Total 515 (delta 73), reused 300 (delta 38), pack-reused 0
Receiving objects: 100% (515/515), 862.32 KiB | 8.89 MiB/s, done.
Resolving deltas: 100% (73/73), done.
Evaluating config script "/home/jenkins/couchdb/src/rebar3/_build/default/lib/rebar/src/rebar.app.src.script"
===> Verifying dependencies...
===> Analyzing applications...
===> Compiling getopt
===> Compiling providers
===> Compiling cf
===> Compiling erlware_commons
===> Compiling bbmustache
===> Compiling relx
===> Compiling eunit_formatters
===> Compiling cth_readable
===> Compiling certifi
===> Compiling ssl_verify_fun
===> Compiling rebar
===> Verifying dependencies...
===> Cleaning out certifi...
===> Cleaning out cf...
===> Cleaning out erlware_commons...
===> Cleaning out getopt...
===> Cleaning out providers...
===> Cleaning out rebar...
===> Verifying dependencies...
===> Analyzing applications...
===> Compiling getopt
===> Compiling providers
===> Compiling cf
===> Compiling erlware_commons
===> Compiling bbmustache
===> Compiling relx
===> Compiling eunit_formatters
===> Compiling cth_readable
===> Compiling certifi
===> Compiling ssl_verify_fun
===> Compiling rebar
===> Building escript for rebar...
Cloning into '/home/jenkins/couchdb/src/erlfmt'...
remote: Enumerating objects: 91, done.
remote: Counting objects: 100% (91/91), done.
remote: Compressing objects: 100% (83/83), done.
remote: Total 91 (delta 6), reused 38 (delta 2), pack-reused 0
Unpacking objects: 100% (91/91), done.
===> Verifying dependencies...
===> Fetching getopt v1.0.2
===> Analyzing applications...
===> Compiling getopt
===> Analyzing applications...
^Z
[1]+  Stopped                 ./configure --spidermonkey-version 78
jenkins@401db7b11852:~/couchdb$ fg
./configure --spidermonkey-version 78
===> Compiling erlfmt
===> Building escript for erlfmt...
===> Verifying dependencies...
===> Cleaning out erlfmt...
==> updating dependencies
WARN:  Expected /home/jenkins/couchdb/src/fauxton to be a raw dependency directory, but no directory found.
==> couch_epi (get-deps)
==> config (get-deps)
==> couch_log (get-deps)
==> khash (get-deps)
WARN:  Missing plugins: [pc]
==> b64url (get-deps)
==> ets_lru (get-deps)
==> chttpd (get-deps)
==> couch (get-deps)
==> couch_event (get-deps)
==> mem3 (get-deps)
==> couch_index (get-deps)
==> couch_mrview (get-deps)
==> couch_replicator (get-deps)
==> couch_plugins (get-deps)
==> couch_pse_tests (get-deps)
==> couch_stats (get-deps)
==> couch_peruser (get-deps)
==> couch_tests (get-deps)
==> couch_dist (get-deps)
==> custodian (get-deps)
==> ddoc_cache (get-deps)
==> dreyfus (get-deps)
==> fabric (get-deps)
==> global_changes (get-deps)
==> ioq (get-deps)
==> jwtf (get-deps)
==> ken (get-deps)
==> mango (get-deps)
==> rexi (get-deps)
==> setup (get-deps)
==> smoosh (get-deps)
==> weatherreport (get-deps)
==> couch_prometheus (get-deps)
==> rel (get-deps)
==> couchdb (get-deps)
WARN:  Expected /home/jenkins/couchdb/src/fauxton to be a raw dependency directory, but no directory found.
Pulling snappy from {git,"https://github.com/apache/couchdb-snappy.git",
                         {tag,"CouchDB-1.0.8"}}
Cloning into 'snappy'...
Pulling fauxton from {git,"https://github.com/apache/couchdb-fauxton",
                          {tag,"v1.2.8"}}
Cloning into 'fauxton'...
Pulling folsom from {git,"https://github.com/apache/couchdb-folsom.git",
                         {tag,"CouchDB-0.8.4"}}
Cloning into 'folsom'...
Pulling hyper from {git,"https://github.com/apache/couchdb-hyper.git",
                        {tag,"CouchDB-2.2.0-7"}}
Cloning into 'hyper'...
Pulling ibrowse from {git,"https://github.com/apache/couchdb-ibrowse.git",
                          {tag,"CouchDB-4.4.2-5"}}
Cloning into 'ibrowse'...
Pulling jiffy from {git,"https://github.com/apache/couchdb-jiffy.git",
                        {tag,"1.1.1"}}
Cloning into 'jiffy'...
Pulling mochiweb from {git,"https://github.com/apache/couchdb-mochiweb.git",
                           {tag,"v3.1.1"}}
Cloning into 'mochiweb'...
Pulling meck from {git,"https://github.com/apache/couchdb-meck.git",
                       {tag,"0.9.2"}}
Cloning into 'meck'...
Pulling recon from {git,"https://github.com/apache/couchdb-recon.git",
                        {tag,"2.5.2"}}
Cloning into 'recon'...
Pulling proper from {git,"https://github.com/proper-testing/proper",
                         {tag,"v1.4"}}
Cloning into 'proper'...
==> snappy (get-deps)
==> fauxton (get-deps)
==> meck (get-deps)
==> folsom (get-deps)
Pulling bear from {git,"https://github.com/apache/couchdb-bear.git",
                       "008f48aff819126e281d5ccae80a258bf9bf9c30"}
Cloning into 'bear'...
==> bear (get-deps)
==> hyper (get-deps)
==> ibrowse (get-deps)
==> jiffy (get-deps)
==> mochiweb (get-deps)
==> recon (get-deps)
WARN:  Missing plugins: [covertool]
==> proper (get-deps)
==> couchdb (update-deps)
Updating snappy from {git,"https://github.com/apache/couchdb-snappy.git",
                          {tag,"CouchDB-1.0.8"}}
Updating fauxton from {git,"https://github.com/apache/couchdb-fauxton",
                           {tag,"v1.2.8"}}
Updating folsom from {git,"https://github.com/apache/couchdb-folsom.git",
                          {tag,"CouchDB-0.8.4"}}
Updating hyper from {git,"https://github.com/apache/couchdb-hyper.git",
                         {tag,"CouchDB-2.2.0-7"}}
Updating ibrowse from {git,"https://github.com/apache/couchdb-ibrowse.git",
                           {tag,"CouchDB-4.4.2-5"}}
Updating jiffy from {git,"https://github.com/apache/couchdb-jiffy.git",
                         {tag,"1.1.1"}}
Updating mochiweb from {git,"https://github.com/apache/couchdb-mochiweb.git",
                            {tag,"v3.1.1"}}
Updating meck from {git,"https://github.com/apache/couchdb-meck.git",
                        {tag,"0.9.2"}}
Updating recon from {git,"https://github.com/apache/couchdb-recon.git",
                         {tag,"2.5.2"}}
Updating proper from {git,"https://github.com/proper-testing/proper",
                          {tag,"v1.4"}}
Updating bear from {git,"https://github.com/apache/couchdb-bear.git",
                        "008f48aff819126e281d5ccae80a258bf9bf9c30"}
WARN:  Missing plugins: [covertool]
You have configured Apache CouchDB, time to relax. Relax.
jenkins@401db7b11852:~/couchdb$ 
jenkins@401db7b11852:~/couchdb$ 
jenkins@401db7b11852:~/couchdb$ 
jenkins@401db7b11852:~/couchdb$ 
jenkins@401db7b11852:~/couchdb$ 
jenkins@401db7b11852:~/couchdb$ make release
fatal: No names found, cannot describe anything.
==> snappy (compile)
Compiled src/snappy.erl
Compiling c_src/snappy_nif.cc
c_src/snappy_nif.cc: In function 'ERL_NIF_TERM snappy_compress(ErlNifEnv*, int, const ERL_NIF_TERM*)':
c_src/snappy_nif.cc:156:28: warning: catching polymorphic type 'class std::bad_alloc' by value [-Wcatch-value=]
     } catch(std::bad_alloc e) {
                            ^
Compiling c_src/snappy/snappy-sinksource.cc
Compiling c_src/snappy/snappy-stubs-internal.cc
Compiling c_src/snappy/snappy.cc
c_src/snappy/snappy.cc: In instantiation of 'bool snappy::SnappyScatteredWriter<Allocator>::AppendFromSelf(size_t, size_t) [with Allocator = snappy::SnappySinkAllocator; size_t = long unsigned int]':
c_src/snappy/snappy.cc:779:13:   required from 'void snappy::SnappyDecompressor::DecompressAllTags(Writer*) [with Writer = snappy::SnappyScatteredWriter<snappy::SnappySinkAllocator>]'
c_src/snappy/snappy.cc:865:3:   required from 'bool snappy::InternalUncompressAllTags(snappy::SnappyDecompressor*, Writer*, snappy::uint32) [with Writer = snappy::SnappyScatteredWriter<snappy::SnappySinkAllocator>; snappy::uint32 = unsigned int]'
c_src/snappy/snappy.cc:1549:78:   required from here
c_src/snappy/snappy.cc:1401:21: warning: comparison of integer expressions of different signedness: 'size_t' {aka 'long unsigned int'} and 'long int' [-Wsign-compare]
     if (offset - 1u < op_ptr_ - op_base_) {
         ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
==> bear (compile)
/home/jenkins/couchdb/src/bear/src/bear.erl:28: Warning: export_all flag enabled - all functions will be exported
Compiled src/bear.erl
==> meck (compile)
Compiled src/meck_util.erl
Compiled src/meck_matcher.erl
Compiled src/meck_ret_spec.erl
Compiled src/meck_history.erl
Compiled src/meck_expect.erl
Compiled src/meck_cover.erl
Compiled src/meck_code.erl
Compiled src/meck_code_gen.erl
Compiled src/meck_args_matcher.erl
Compiled src/meck.erl
Compiled src/meck_proc.erl
==> folsom (compile)
Compiled src/folsom_sup.erl
Compiled src/folsom_utils.erl
Compiled src/folsom_sample_uniform.erl
Compiled src/folsom_sample_slide_sup.erl
Compiled src/folsom_sample_slide_uniform.erl
Compiled src/folsom_sample_slide_sorted.erl
Compiled src/folsom_sample_slide_server.erl
Compiled src/folsom_sample_slide.erl
Compiled src/folsom_sample_none.erl
Compiled src/folsom_sample.erl
Compiled src/folsom_metrics_spiral.erl
Compiled src/folsom_sample_exdec.erl
Compiled src/folsom_vm_metrics.erl
Compiled src/folsom_metrics_meter_reader.erl
Compiled src/folsom_metrics_histogram_ets.erl
Compiled src/folsom_metrics_history.erl
Compiled src/folsom_metrics_meter.erl
Compiled src/folsom_metrics_gauge.erl
Compiled src/folsom_metrics_histogram.erl
Compiled src/folsom_metrics_duration.erl
Compiled src/folsom_metrics_counter.erl
Compiled src/folsom_meter_timer_server.erl
Compiled src/folsom_ewma.erl
Compiled src/folsom_metrics.erl
Compiled src/folsom.erl
Compiled src/folsom_ets.erl
==> hyper (compile)
Compiled src/hyper_register.erl
Compiled src/hyper_carray.erl
Compiled src/hyper_const.erl
Compiled src/hyper_gb.erl
Compiled src/hyper_array.erl
Compiled src/hyper_binary.erl
Compiled src/hyper_binary_rle.erl
Compiled src/hyper.erl
Compiling c_src/hyper_carray.c
==> ibrowse (compile)
Compiled src/ibrowse_lib.erl
Compiled src/ibrowse_sup.erl
Compiled src/ibrowse_socks5.erl
Compiled src/ibrowse_app.erl
Compiled src/ibrowse_lb.erl
Compiled src/ibrowse.erl
Compiled src/ibrowse_http_client.erl
==> jiffy (compile)
Compiled src/jiffy_utf8.erl
Compiled src/jiffy.erl
Compiling c_src/decoder.c
Compiling c_src/encoder.c
Compiling c_src/jiffy.c
Compiling c_src/termstack.c
Compiling c_src/utf8.c
Compiling c_src/util.c
Compiling c_src/doubles.cc
Compiling c_src/objects.cc
Compiling c_src/double-conversion/bignum-dtoa.cc
Compiling c_src/double-conversion/bignum.cc
Compiling c_src/double-conversion/cached-powers.cc
Compiling c_src/double-conversion/diy-fp.cc
Compiling c_src/double-conversion/double-conversion.cc
Compiling c_src/double-conversion/fast-dtoa.cc
Compiling c_src/double-conversion/fixed-dtoa.cc
Compiling c_src/double-conversion/strtod.cc
==> mochiweb (compile)
Compiled src/reloader.erl
Compiled src/mochiweb_websocket.erl
Compiled src/mochiweb_util.erl
Compiled src/mochiweb_socket.erl
Compiled src/mochiweb_socket_server.erl
Compiled src/mochiweb_session.erl
Compiled src/mochiweb_response.erl
Compiled src/mochiweb_mime.erl
Compiled src/mochiweb_io.erl
Compiled src/mochiweb_multipart.erl
Compiled src/mochiweb_http.erl
Compiled src/mochiweb_headers.erl
Compiled src/mochiweb_echo.erl
Compiled src/mochiweb_cover.erl
Compiled src/mochiweb_cookies.erl
Compiled src/mochiweb_clock.erl
Compiled src/mochiweb_request.erl
Compiled src/mochiweb_base64url.erl
Compiled src/mochiweb_acceptor.erl
Compiled src/mochiweb.erl
Compiled src/mochiutf8.erl
Compiled src/mochitemp.erl
Compiled src/mochinum.erl
Compiled src/mochilogfile2.erl
Compiled src/mochilists.erl
Compiled src/mochiweb_html.erl
Compiled src/mochijson.erl
Compiled src/mochihex.erl
Compiled src/mochijson2.erl
Compiled src/mochiglobal.erl
Compiled src/mochifmt_std.erl
Compiled src/mochifmt_records.erl
Compiled src/mochifmt.erl
Compiled src/mochiweb_charref.erl
==> recon (compile)
Compiled src/recon_map.erl
Compiled src/recon_rec.erl
Compiled src/recon_trace.erl
Compiled src/recon_lib.erl
Compiled src/recon_alloc.erl
Compiled src/recon.erl
WARN:  Missing plugins: [covertool]
==> proper (compile)
Compiled src/vararg.erl
Compiled src/proper_target.erl
Compiled src/proper_unicode.erl
Compiled src/proper_unused_imports_remover.erl
Compiled src/proper_transformer.erl
Compiled src/proper_symb.erl
Compiled src/proper_types.erl
Compiled src/proper_statem.erl
Compiled src/proper_sets.erl
Compiled src/proper_sa.erl
Compiled src/proper_queue.erl
Compiled src/proper_shrink.erl
Compiled src/proper_ordsets.erl
Compiled src/proper_prop_remover.erl
Compiled src/proper_orddict.erl
Compiled src/proper_gen.erl
Compiled src/proper_gb_trees.erl
Compiled src/proper_gb_sets.erl
Compiled src/proper_gen_next.erl
Compiled src/proper_fsm.erl
Compiled src/proper_dict.erl
Compiled src/proper_array.erl
Compiled src/proper_typeserver.erl
Compiled src/proper_arith.erl
Compiled src/proper.erl
Compiled src/proper_erlang_abstract_code.erl
==> couch_epi (compile)
Compiled src/couch_epi_util.erl
Compiled src/couch_epi_sup.erl
Compiled src/couch_epi_module_keeper.erl
Compiled src/couch_epi_plugin.erl
Compiled src/couch_epi_functions.erl
Compiled src/couch_epi_data_gen.erl
Compiled src/couch_epi_data.erl
Compiled src/couch_epi_codechange_monitor.erl
Compiled src/couch_epi_app.erl
Compiled src/couch_epi_codegen.erl
Compiled src/couch_epi_functions_gen.erl
Compiled src/couch_epi.erl
==> config (compile)
Compiled src/config_sup.erl
Compiled src/config_util.erl
Compiled src/config_notifier.erl
Compiled src/config_writer.erl
Compiled src/config_listener.erl
Compiled src/config_listener_mon.erl
Compiled src/config_app.erl
Compiled src/config.erl
==> couch_log (compile)
Compiled src/couch_log_writer.erl
Compiled src/couch_log_writer_stderr.erl
Compiled src/couch_log_writer_journald.erl
Compiled src/couch_log_writer_syslog.erl
Compiled src/couch_log_writer_file.erl
Compiled src/couch_log_util.erl
Compiled src/couch_log_sup.erl
Compiled src/couch_log_server.erl
Compiled src/couch_log_monitor.erl
Compiled src/couch_log_trunc_io_fmt.erl
Compiled src/couch_log_error_logger_h.erl
Compiled src/couch_log_config_dyn.erl
Compiled src/couch_log_config.erl
Compiled src/couch_log_app.erl
Compiled src/couch_log.erl
Compiled src/couch_log_formatter.erl
Compiled src/couch_log_trunc_io.erl
==> khash (compile)
Compiled src/khash.erl
Compiling c_src/hash.c
Compiling c_src/khash.c
WARN:  Missing plugins: [pc]
==> b64url (compile)
Compiled src/b64url.erl
Compiling c_src/b64url.c
==> ets_lru (compile)
Compiled src/ets_lru.erl
==> chttpd (compile)
Compiled src/chttpd.erl
Compiled src/chttpd_util.erl
Compiled src/chttpd_xframe_options.erl
Compiled src/chttpd_test_util.erl
Compiled src/chttpd_stats.erl
Compiled src/chttpd_view.erl
Compiled src/chttpd_sup.erl
Compiled src/chttpd_prefer_header.erl
Compiled src/chttpd_plugin.erl
Compiled src/chttpd_show.erl
Compiled src/chttpd_rewrite.erl
Compiled src/chttpd_httpd_handlers.erl
Compiled src/chttpd_handlers.erl
Compiled src/chttpd_misc.erl
Compiled src/chttpd_node.erl
Compiled src/chttpd_epi.erl
Compiled src/chttpd_cors.erl
Compiled src/chttpd_external.erl
Compiled src/chttpd_auth_request.erl
Compiled src/chttpd_auth.erl
Compiled src/chttpd_app.erl
Compiled src/chttpd_auth_cache.erl
Compiled src/chttpd_db.erl
==> couch (compile)
Compiled src/couch_httpd.erl
Compiled src/test_request.erl
Compiled src/couch_uuids.erl
Compiled src/couch_work_queue.erl
Compiled src/couch_users_db.erl
Compiled src/couch_totp.erl
Compiled src/test_util.erl
Compiled src/couch_task_status.erl
Compiled src/couch_sup.erl
Compiled src/couch_stream.erl
Compiled src/couch_secondary_sup.erl
Compiled src/couch_rand.erl
Compiled src/couch_util.erl
Compiled src/couch_proc_manager.erl
Compiled src/couch_query_servers.erl
Compiled src/couch_server.erl
Compiled src/couch_primary_sup.erl
Compiled src/couch_password_hasher.erl
Compiled src/couch_partition.erl
Compiled src/couch_passwords.erl
Compiled src/couch_os_process.erl
Compiled src/couch_lru.erl
Compiled src/couch_multidb_changes.erl
Compiled src/couch_io_logger.erl
Compiled src/couch_native_process.erl
Compiled src/couch_httpd_vhost.erl
Compiled src/couch_key_tree.erl
Compiled src/couch_httpd_rewrite.erl
Compiled src/couch_httpd_handlers.erl
Compiled src/couch_httpd_misc_handlers.erl
Compiled src/couch_httpd_multipart.erl
Compiled src/couch_hotp.erl
Compiled src/couch_hash.erl
Compiled src/couch_flags_config.erl
Compiled src/couch_flags.erl
Compiled src/couch_httpd_auth.erl
Compiled src/couch_event_sup.erl
Compiled src/couch_emsort.erl
Compiled src/couch_ejson_size.erl
Compiled src/couch_ejson_compare.erl
Compiled src/couch_file.erl
Compiled src/couch_httpd_db.erl
Compiled src/couch_doc.erl
Compiled src/couch_debug.erl
Compiled src/couch_db_plugin.erl
Compiled src/couch_db_header.erl
Compiled src/couch_db_epi.erl
Compiled src/couch_db_split.erl
Compiled src/couch_db_engine.erl
Compiled src/couch_db_updater.erl
Compiled src/couch_compress.erl
Compiled src/couch_changes.erl
Compiled src/couch_bt_engine_stream.erl
Compiled src/couch_bt_engine_header.erl
Compiled src/couch_btree.erl
Compiled src/couch_bt_engine_compactor.erl
Compiled src/couch_db.erl
Compiled src/couch_base32.erl
Compiled src/couch_auth_cache.erl
Compiled src/couch_app.erl
Compiled src/couch.erl
Compiled src/couch_bt_engine.erl
Compiled src/couch_att.erl
Compiling /home/jenkins/couchdb/src/couch/priv/couch_js/86/main.cpp
/home/jenkins/couchdb/src/couch/priv/couch_js/86/main.cpp:24:10: fatal error: jsapi.h: No such file or directory
 #include <jsapi.h>
          ^~~~~~~~~
compilation terminated.
ERROR: compile failed while processing /home/jenkins/couchdb/src/couch: rebar_abort
make: *** [Makefile:123: couch] Error 1
jenkins@401db7b11852:~/couchdb$ echo $?
2
```

I could try rocky-8/ubi8 instead, not sure if it works there or has maybe a different default spidermonkey version available.
But `--spidermonkey-version 78` seem to be not working on bullseye.

--- a/.gitignore
+++ b/.gitignore
@@ -38,6 +38,7 @@ rel/dev*
 rel/tmpdata
 share/server/main-coffee.js
 share/server/main.js
+share/server/main-ast-bypass.js
 share/www
 src/bear/
 src/certifi/

--- a/Makefile
+++ b/Makefile
@@ -442,7 +442,7 @@ clean:
 	@rm -rf src/*/.rebar
 	@rm -rf src/*/priv/*.so
 	@rm -rf src/couch/priv/{couchspawnkillable,couchjs}
-	@rm -rf share/server/main.js share/server/main_ast_bypass.js share/server/main-coffee.js
+	@rm -rf share/server/main.js share/server/main-ast-bypass.js share/server/main-coffee.js
 	@rm -rf tmp dev/data dev/lib dev/logs
 	@rm -rf src/mango/.venv
 	@rm -f src/couch/priv/couchspawnkillable

--- a/Makefile
+++ b/Makefile
@@ -442,7 +442,7 @@ clean:
 	@rm -rf src/*/.rebar
 	@rm -rf src/*/priv/*.so
 	@rm -rf src/couch/priv/{couchspawnkillable,couchjs}
-	@rm -rf share/server/main.js share/server/main-coffee.js
+	@rm -rf share/server/main.js share/server/main_ast_bypass.js share/server/main-coffee.js
 	@rm -rf tmp dev/data dev/lib dev/logs
 	@rm -rf src/mango/.venv
 	@rm -f src/couch/priv/couchspawnkillable

--- a/Makefile.win
+++ b/Makefile.win
@@ -374,7 +374,7 @@ clean:
 	-@rmdir /s/q src\*\.rebar
 	-@del /f/q/s src\*.dll
 	-@del /f/q src\couch\priv\*.exe
-	-@del /f/q share\server\main.js share\server\main-coffee.js
+	-@del /f/q share\server\main.js share\server\main_ast_bypass.js share\server\main-coffee.js
 	-@rmdir /s/q tmp
 	-@rmdir /s/q dev\data
 	-@rmdir /s/q dev\lib

--- a/Makefile.win
+++ b/Makefile.win
@@ -374,7 +374,7 @@ clean:
 	-@rmdir /s/q src\*\.rebar
 	-@del /f/q/s src\*.dll
 	-@del /f/q src\couch\priv\*.exe
-	-@del /f/q share\server\main.js share\server\main_ast_bypass.js share\server\main-coffee.js
+	-@del /f/q share\server\main.js share\server\main-ast-bypass.js share\server\main-coffee.js
 	-@rmdir /s/q tmp
 	-@rmdir /s/q dev\data
 	-@rmdir /s/q dev\lib

--- a/rel/reltool.config
+++ b/rel/reltool.config
@@ -141,6 +141,7 @@
     {copy, "overlay/etc"},
     {copy, "../src/couch/priv/couchjs", "bin/couchjs"},
     {copy, "../share/server/main.js", "share/server/main.js"},
+    {copy, "../share/server/main-ast-bypass.js", "share/server/main-ast-bypass.js"},
     {copy, "../share/server/main-coffee.js", "share/server/main-coffee.js"},
     {copy, "../src/weatherreport/weatherreport", "bin/weatherreport"},
     {copy, "files/sys.config", "releases/\{\{rel_vsn\}\}/sys.config"},

--- a/share/server/60/rewrite_fun.js
+++ b/share/server/60/rewrite_fun.js
@@ -16,6 +16,12 @@
 //  https://github.com/dmunch/couch-chakra/blob/master/js/normalizeFunction.js
 
 function rewriteFunInt(fun) {
+    // Skip lengthy AST transforms if the function passed can be
+    // safely wrapped in parenthesis.
+    if (fun.startsWith('function') && fun.endsWith('}')) {
+        return '(' + fun + ')'
+    }
+
     const ast = esprima.parse(fun);
     let idx = ast.body.length - 1;
     let decl = {};

--- a/share/server/60/rewrite_fun_ast_bypass.js
+++ b/share/server/60/rewrite_fun_ast_bypass.js
@@ -16,6 +16,12 @@
 //  https://github.com/dmunch/couch-chakra/blob/master/js/normalizeFunction.js
 
 function rewriteFunInt(fun) {
+    // Skip lengthy AST transforms if the function passed can be
+    // safely wrapped in parentheses.
+    if (fun.startsWith('function') && fun.endsWith('}')) {
+        return '(' + fun + ')'
+    }
+
     const ast = esprima.parse(fun);
     let idx = ast.body.length - 1;
     let decl = {};

--- a/src/docs/src/config/query-servers.rst
+++ b/src/docs/src/config/query-servers.rst
@@ -65,6 +65,19 @@ you can adjust the memory limitation (here, increasing to 512 MiB)::
 
 For more info about the available options, please consult ``couchjs -h``.
 
+.. note::
+    CouchDB versions 3.0.0 to 3.2.2 included a performance regression for
+    custom reduce functions. CouchDB 3.3.0 and later come with an experimental
+    fix to this issue that is included in a separate ``.js`` file.
+
+    To enable the fix, you need to define a custom ``COUCHDB_QUERY_SERVER_JAVASCRIPT``
+    environment variable as outlined above. The path to ``couchjs`` needs to
+    remain the same as you find it on your ``couchdb`` file, and the path to
+    ``main.js`` needs to be set to ``/path/to/couchdb/share/server/main-ast-bypass.js``.
+
+    With a default installation on Linux systems, this is going to be
+    ``COUCHDB_QUERY_SERVER_JAVASCRIPT="/opt/couchdb/bin/couchjs /opt/couchdb/share/server/main-ast-bypass.js"``
+
 .. _Mozilla SpiderMonkey: https://spidermonkey.dev/
 
 .. seealso::

--- a/support/build_js.escript
+++ b/support/build_js.escript
@@ -90,6 +90,7 @@ main([]) ->
     case SMVsn of
         "1.8.5" ->
             ok = Concat(ExtraFiles ++ JsFiles, "share/server/main.js");
+            ok = Concat(ExtraFiles ++ JsFiles, "share/server/main-ast-bypass.js");
         _ ->
             ok = Concat(RewriteFunFile ++ ExtraFiles ++ JsFiles, "share/server/main.js"),
             ok = Concat(RewriteFunWithASTBypassFile ++ ExtraFiles ++ JsFiles, "share/server/main-ast-bypass.js")

--- a/support/build_js.escript
+++ b/support/build_js.escript
@@ -68,10 +68,12 @@ main([]) ->
         _ ->
             [
                 "share/server/60/esprima.js",
-                "share/server/60/escodegen.js",
-                "share/server/60/rewrite_fun.js"
+                "share/server/60/escodegen.js"
             ]
     end,
+
+    RewriteFunFile = ["share/server/60/rewrite_fun.js"],
+    RewriteFunWithASTBypassFile = ["share/server/60/rewrite_fun_ast_bypass.js"],
 
     Pre = "(function () {\n",
     Post = "})();\n",
@@ -85,6 +87,12 @@ main([]) ->
             file:write_file(To, FinalBin)
     end,
 
-    ok = Concat(ExtraFiles ++ JsFiles, "share/server/main.js"),
+    case SMVsn of
+        "1.8.5" ->
+            ok = Concat(ExtraFiles ++ JsFiles, "share/server/main.js");
+        _ ->
+            ok = Concat(RewriteFunFile ++ ExtraFiles ++ JsFiles, "share/server/main.js"),
+            ok = Concat(RewriteFunWithASTBypassFile ++ ExtraFiles ++ JsFiles, "share/server/main_ast_bypass.js")
+    end,
     ok = Concat(ExtraFiles ++ CoffeeFiles, "share/server/main-coffee.js"),
     ok.

--- a/support/build_js.escript
+++ b/support/build_js.escript
@@ -89,7 +89,7 @@ main([]) ->
 
     case SMVsn of
         "1.8.5" ->
-            ok = Concat(ExtraFiles ++ JsFiles, "share/server/main.js");
+            ok = Concat(ExtraFiles ++ JsFiles, "share/server/main.js"),
             ok = Concat(ExtraFiles ++ JsFiles, "share/server/main-ast-bypass.js");
         _ ->
             ok = Concat(RewriteFunFile ++ ExtraFiles ++ JsFiles, "share/server/main.js"),

--- a/support/build_js.escript
+++ b/support/build_js.escript
@@ -92,7 +92,7 @@ main([]) ->
             ok = Concat(ExtraFiles ++ JsFiles, "share/server/main.js");
         _ ->
             ok = Concat(RewriteFunFile ++ ExtraFiles ++ JsFiles, "share/server/main.js"),
-            ok = Concat(RewriteFunWithASTBypassFile ++ ExtraFiles ++ JsFiles, "share/server/main_ast_bypass.js")
+            ok = Concat(RewriteFunWithASTBypassFile ++ ExtraFiles ++ JsFiles, "share/server/main-ast-bypass.js")
     end,
     ok = Concat(ExtraFiles ++ CoffeeFiles, "share/server/main-coffee.js"),
     ok.

